### DISCUSSION TITLE
Python Type Hinting

### DISCUSSION THREAD

**Discussion by KieranBrannigan** [rdkit/rdkit#3981] [Q&A]
I have installed rdkit for python3.6 on Windows using conda-forge. When I am writing some python code in VSCode, it would be very useful to have some type hinting available. Currently, for example I even get imports not being recognised when I import from rdkit.Chem for example:
`from rdkit.Chem import Mol, SanitizeFlags, SanitizeMol, MolFromSmiles, RDKFingerprint` all of these imports are not recognised.

Also when writing code the imported functions and classes have no type hinting available.

I'm wondering whether type hinting should be available and I have installed something incorrectly, or whether type hinting could be installed through conda.

**Comment by greglandrum** [comment]
The RDKit currently does not provide type hinting.

**Comment by bertiewooster** [comment]
Would it be useful for contributors to provide type hinting for new Python code?

Related question: which versions of Python does the RDKit support? (I couldn't readily find that info in the docs.) I ask because [different versions of Python support different levels of type hints](https://stackoverflow.com/questions/31905597/how-to-properly-function-annotate-type-hint-a-list-of-strings#31905868).

**Reply by greglandrum** [reply]
I find that type hints make Python code less readable. If there are areas where it's unclear what type an argument should be, that should be handled in the doc string.

We don't capture anywhere which versions of Python are supported. We should try to support all officially supported versions of Python (or at least whatever the conda-forge team does binaries for). I think at the moment it's back to 3.7 and up to 3.11.

**Reply by bertiewooster** [reply]
OK, thanks.

**Reply by santi-mir** [reply]
Hi!

Is there the possibility to import `Mol` so that we can define our own types to rdkit functions @greglandrum ? 

Like `m:Mol = rdkit.MolFromSmiles(smile)`

This is actually quite useful for the users of the code, not just the devs.

I'm also curious as to whether there is a way to avoid this error when importing?

`"MolFromSmiles" is not a known attribute of module "rdkit.Chem"`

**Reply by bertiewooster** [reply]
Regarding your first question, either `Chem.Mol` or `Chem.rdchem.Mol` works as a type for me:

```
from rdkit import Chem

smile:str = "C1=CC=CC=C1"
m:Chem.Mol = Chem.MolFromSmiles(smile)
print(f"{m=}")
m2:Chem.rdchem.Mol = Chem.MolFromSmiles(smile)
print(f"{m2=}")
canon_smile:str = Chem.MolToSmiles(m)
print(f"{canon_smile=}")
canon_smile2:str = Chem.MolToSmiles(m2)
print(f"{canon_smile2=}")
```

Output:
```
m=<rdkit.Chem.rdchem.Mol object at 0x114638120>
m2=<rdkit.Chem.rdchem.Mol object at 0x114638190>
canon_smile='c1ccccc1'
canon_smile2='c1ccccc1'
```

From the Mol object outputs, `Chem.rdchem.Mol` seems to be the recommended type, but `Chem.Mol` works too and is more concise.

**Reply by santi-mir** [reply]
Thank you!
What do you get if you hover on "Mol" in vscode? I see "Unknown"

**Reply by bertiewooster** [reply]
VSCode shows me `(function) Mol: Any` on hovering over either Mol type hint.

**Reply by jones-gareth** [reply]
I see the doc for Mol when hovering over the type hint.  This may be to do with using the `Jedi` language server.  I've found that `Jedi` gives better intellisense for RDKit objects compared to the default `Pylance`.

**Reply by santi-mir** [reply]
Moved to Jedi, it's great thank you @jones-gareth

**Comment by bionicles** [comment]
Hey I could use type hints right now, this plus formatting the code with flake8 or black could be an easy massive upgrade for RDKit, I'll open an issue 

Edit: Opened https://github.com/rdkit/rdkit/issues/7682#issue-2442477807
